File: cmd/kavach/internal/commands/quality/quality.go
// Package quality provides standalone quality commands.
// quality.go: Code quality analysis with SP/1.0 TOON output.
// P3 FIX: Standalone quality command for direct CLI usage.
package quality

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/claude/shared/pkg/util"
	"github.com/claude/shared/pkg/validate"
	"github.com/spf13/cobra"
)

var (
	qualityFormat  string
	qualityVerbose bool
)

// QualityCmd is the top-level quality command.
var QualityCmd = &cobra.Command{
	Use:   "quality [file|dir...]",
	Short: "Analyze code quality metrics",
	Long: `[QUALITY]
desc: Analyze code quality and complexity metrics
output: SP/1.0 TOON format

[METRICS]
lines:        Total lines of code
functions:    Function count and complexity
dace_score:   DACE micro-modular compliance (0-100)
ast_valid:    AST syntax validation
imports:      Import analysis

[USAGE]
kavach quality file.go          # Analyze single file
kavach quality src/             # Analyze directory
kavach quality --format=json    # JSON output
kavach quality --verbose        # Detailed output`,
	Run: runQuality,
}

func init() {
	QualityCmd.Flags().StringVar(&qualityFormat, "format", "toon", "Output format (toon, json)")
	QualityCmd.Flags().BoolVar(&qualityVerbose, "verbose", false, "Verbose output")
}

// QualityResult represents quality metrics for a file.
type QualityResult struct {
	File       string `json:"file"`
	Lines      int    `json:"lines"`
	Functions  int    `json:"functions"`
	Imports    int    `json:"imports"`
	DACEScore  int    `json:"dace_score"`
	ASTValid   bool   `json:"ast_valid"`
	ASTError   string `json:"ast_error,omitempty"`
	Complexity string `json:"complexity"`
}

func runQuality(cmd *cobra.Command, args []string) {
	if len(args) == 0 {
		cmd.Help()
		return
	}

	var results []QualityResult

	for _, arg := range args {
		info, err := os.Stat(arg)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error: %s: %v\n", arg, err)
			continue
		}

		if info.IsDir() {
			filepath.Walk(arg, func(path string, info os.FileInfo, err error) error {
				if err != nil || info.IsDir() {
					return nil
				}
				ext := util.GetExtension(path)
				if isAnalyzable(ext) {
					result := analyzeFile(path)
					results = append(results, result)
				}
				return nil
			})
		} else {
			result := analyzeFile(arg)
			results = append(results, result)
		}
	}

	outputQualityResults(results)
}

func isAnalyzable(ext string) bool {
	analyzable := map[string]bool{
		".go": true, ".rs": true, ".ts": true, ".tsx": true,
		".js": true, ".jsx": true, ".py": true,
	}
	return analyzable[ext]
}

func analyzeFile(path string) QualityResult {
	result := QualityResult{
		File:     path,
		ASTValid: true,
	}

	content, err := os.ReadFile(path)
	if err != nil {
		result.ASTValid = false
		result.ASTError = err.Error()
		return result
	}

	lines := strings.Split(string(content), "\n")
	result.Lines = len(lines)

	ext := util.GetExtension(path)

	// Count functions (simple heuristic)
	result.Functions = countFunctions(string(content), ext)

	// Count imports
	result.Imports = countImports(string(content), ext)

	// AST validation
	switch ext {
	case ".go":
		if errMsg := validate.GoSyntax(string(content)); errMsg != "" {
			result.ASTValid = false
			result.ASTError = errMsg
		}
	case ".json":
		if errMsg := validate.JSONSyntax(string(content)); errMsg != "" {
			result.ASTValid = false
			result.ASTError = errMsg
		}
	}

	// Calculate DACE score (0-100)
	result.DACEScore = calculateDACEScore(result)

	// Determine complexity
	result.Complexity = determineComplexity(result)

	return result
}

func countFunctions(content, ext string) int {
	count := 0
	lines := strings.Split(content, "\n")

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		switch ext {
		case ".go":
			if strings.HasPrefix(trimmed, "func ") {
				count++
			}
		case ".rs":
			if strings.HasPrefix(trimmed, "fn ") || strings.HasPrefix(trimmed, "pub fn ") {
				count++
			}
		case ".ts", ".tsx", ".js", ".jsx":
			if strings.Contains(trimmed, "function ") ||
				strings.Contains(trimmed, "=> {") ||
				strings.Contains(trimmed, "async ") {
				count++
			}
		case ".py":
			if strings.HasPrefix(trimmed, "def ") || strings.HasPrefix(trimmed, "async def ") {
				count++
			}
		}
	}

	return count
}

func countImports(content, ext string) int {
	count := 0
	lines := strings.Split(content, "\n")

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		switch ext {
		case ".go":
			if strings.HasPrefix(trimmed, "import ") || trimmed == "import (" {
				count++
			}
			if strings.HasPrefix(trimmed, "\"") && strings.HasSuffix(trimmed, "\"") {
				count++
			}
		case ".rs":
			if strings.HasPrefix(trimmed, "use ") {
				count++
			}
		case ".ts", ".tsx", ".js", ".jsx":
			if strings.HasPrefix(trimmed, "import ") || strings.HasPrefix(trimmed, "require(") {
				count++
			}
		case ".py":
			if strings.HasPrefix(trimmed, "import ") || strings.HasPrefix(trimmed, "from ") {
				count++
			}
		}
	}

	return count
}

func calculateDACEScore(r QualityResult) int {
	score := 100

	// Deduct for lines over 100 (DACE micro-modular)
	if r.Lines > 100 {
		deduct := (r.Lines - 100) / 10
		if deduct > 50 {
			deduct = 50
		}
		score -= deduct
	}

	// Deduct for too many functions (suggests file should be split)
	if r.Functions > 10 {
		deduct := (r.Functions - 10) * 2
		if deduct > 20 {
			deduct = 20
		}
		score -= deduct
	}

	// Deduct for AST errors
	if !r.ASTValid {
		score -= 30
	}

	if score < 0 {
		score = 0
	}

	return score
}

func determineComplexity(r QualityResult) string {
	if r.Lines <= 50 && r.Functions <= 5 {
		return "low"
	}
	if r.Lines <= 100 && r.Functions <= 10 {
		return "medium"
	}
	return "high"
}

func outputQualityResults(results []QualityResult) {
	if qualityFormat == "json" {
		fmt.Println("[")
		for i, r := range results {
			fmt.Printf("  {\"file\": %q, \"lines\": %d, \"dace_score\": %d, \"ast_valid\": %v}",
				r.File, r.Lines, r.DACEScore, r.ASTValid)
			if i < len(results)-1 {
				fmt.Println(",")
			} else {
				fmt.Println()
			}
		}
		fmt.Println("]")
		return
	}

	// TOON output (default)
	totalLines := 0
	totalFunctions := 0
	avgDACE := 0
	for _, r := range results {
		totalLines += r.Lines
		totalFunctions += r.Functions
		avgDACE += r.DACEScore
	}
	if len(results) > 0 {
		avgDACE /= len(results)
	}

	fmt.Println("[QUALITY_SUMMARY]")
	fmt.Printf("files: %d\n", len(results))
	fmt.Printf("total_lines: %d\n", totalLines)
	fmt.Printf("total_functions: %d\n", totalFunctions)
	fmt.Printf("avg_dace_score: %d\n", avgDACE)
	fmt.Println()

	if qualityVerbose {
		for _, r := range results {
			fmt.Printf("[FILE:%s]\n", r.File)
			fmt.Printf("  lines: %d\n", r.Lines)
			fmt.Printf("  functions: %d\n", r.Functions)
			fmt.Printf("  imports: %d\n", r.Imports)
			fmt.Printf("  dace_score: %d\n", r.DACEScore)
			fmt.Printf("  complexity: %s\n", r.Complexity)
			if !r.ASTValid {
				fmt.Printf("  ast_error: %s\n", r.ASTError)
			}
			fmt.Println()
		}
	}
}

File: crates/kavach-cli/src/commands/quality.rs
//! Quality: code quality analysis with DACE scoring.
//! Metrics: lines, functions, imports, DACE score, complexity.
//! Output: TOON (default) or JSON format.

use std::io::Write;
use std::path::Path;

use clap::Args;

#[derive(Args)]
pub struct QualityArgs {
    #[arg(long)]
    pub format: Option<String>,
    #[arg(long)]
    pub verbose: bool,
    pub paths: Vec<String>,
}

struct QualityResult {
    file: String,
    lines: usize,
    functions: usize,
    imports: usize,
    dace_score: i32,
    complexity: String,
}

pub fn run(args: QualityArgs) -> anyhow::Result<()> {
    if args.paths.is_empty() {
        crate::commands::cli_print("Usage: kavach quality [--verbose] [--format=toon|json] <file|dir>...");
        return Ok(());
    }

    let mut results = Vec::new();

    for arg in &args.paths {
        let path = Path::new(arg);
        if !path.exists() {
            let stderr = std::io::stderr();
            let mut h = stderr.lock();
            let _ = writeln!(h, "error: {arg}: not found");
            continue;
        }

        if path.is_dir() {
            walk_dir(path, &mut results);
        } else {
            results.push(analyze_file(path));
        }
    }

    let fmt = args.format.as_deref().unwrap_or("toon");
    if fmt == "json" {
        output_json(&results)
    } else {
        output_toon(&results, args.verbose)
    }
}

fn walk_dir(dir: &Path, results: &mut Vec<QualityResult>) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            let name = path.file_name().unwrap_or_default().to_string_lossy();
            if name.starts_with('.') || name == "target" || name == "node_modules" {
                continue;
            }
            walk_dir(&path, results);
        } else if is_analyzable(&path) {
            results.push(analyze_file(&path));
        }
    }
}

fn is_analyzable(path: &Path) -> bool {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
    matches!(ext, "go" | "rs" | "ts" | "tsx" | "js" | "jsx" | "py")
}

fn analyze_file(path: &Path) -> QualityResult {
    let file_str = path.to_string_lossy().to_string();
    let content = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => {
            return QualityResult {
                file: file_str, lines: 0, functions: 0,
                imports: 0, dace_score: 0, complexity: "unknown".into(),
            };
        }
    };

    let lines: Vec<&str> = content.lines().collect();
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");

    let functions = count_functions(&lines, ext);
    let imports = count_imports(&lines, ext);

    let mut result = QualityResult {
        file: file_str,
        lines: lines.len(),
        functions,
        imports,
        dace_score: 100,
        complexity: String::new(),
    };

    // DACE score
    if result.lines > 100 {
        let deduct = ((result.lines - 100) / 10).min(50) as i32;
        result.dace_score -= deduct;
    }
    if result.functions > 10 {
        let deduct = ((result.functions - 10) * 2).min(20) as i32;
        result.dace_score -= deduct;
    }
    if result.dace_score < 0 { result.dace_score = 0; }

    // Complexity
    result.complexity = if result.lines <= 50 && result.functions <= 5 {
        "low".into()
    } else if result.lines <= 100 && result.functions <= 10 {
        "medium".into()
    } else {
        "high".into()
    };

    result
}

fn count_functions(lines: &[&str], ext: &str) -> usize {
    let mut count = 0;
    for line in lines {
        let trimmed = line.trim();
        match ext {
            "go" => {
                if trimmed.starts_with("func ") { count += 1; }
            }
            "rs" => {
                if trimmed.starts_with("fn ") || trimmed.starts_with("pub fn ") { count += 1; }
            }
            "ts" | "tsx" | "js" | "jsx" => {
                if trimmed.contains("function ") || trimmed.contains("=> {") || trimmed.contains("async ") {
                    count += 1;
                }
            }
            "py" => {
                if trimmed.starts_with("def ") || trimmed.starts_with("async def ") { count += 1; }
            }
            _ => {}
        }
    }
    count
}

fn count_imports(lines: &[&str], ext: &str) -> usize {
    let mut count = 0;
    for line in lines {
        let trimmed = line.trim();
        match ext {
            "go" => {
                if trimmed.starts_with("import ") || trimmed == "import (" { count += 1; }
                if trimmed.starts_with('"') && trimmed.ends_with('"') { count += 1; }
            }
            "rs" => {
                if trimmed.starts_with("use ") { count += 1; }
            }
            "ts" | "tsx" | "js" | "jsx" => {
                if trimmed.starts_with("import ") || trimmed.starts_with("require(") { count += 1; }
            }
            "py" => {
                if trimmed.starts_with("import ") || trimmed.starts_with("from ") { count += 1; }
            }
            _ => {}
        }
    }
    count
}

fn output_toon(results: &[QualityResult], verbose: bool) -> anyhow::Result<()> {
    let out = std::io::stdout();
    let mut w = out.lock();

    let total_lines: usize = results.iter().map(|r| r.lines).sum();
    let total_functions: usize = results.iter().map(|r| r.functions).sum();
    let avg_dace = if results.is_empty() { 0 } else {
        results.iter().map(|r| r.dace_score).sum::<i32>() / results.len() as i32
    };

    writeln!(w, "[QUALITY_SUMMARY]")?;
    writeln!(w, "files: {}", results.len())?;
    writeln!(w, "total_lines: {total_lines}")?;
    writeln!(w, "total_functions: {total_functions}")?;
    writeln!(w, "avg_dace_score: {avg_dace}")?;
    writeln!(w)?;

    if verbose {
        for r in results {
            writeln!(w, "[FILE:{}]", r.file)?;
            writeln!(w, "  lines: {}", r.lines)?;
            writeln!(w, "  functions: {}", r.functions)?;
            writeln!(w, "  imports: {}", r.imports)?;
            writeln!(w, "  dace_score: {}", r.dace_score)?;
            writeln!(w, "  complexity: {}", r.complexity)?;
            writeln!(w)?;
        }
    }

    Ok(())
}

fn output_json(results: &[QualityResult]) -> anyhow::Result<()> {
    let out = std::io::stdout();
    let mut w = out.lock();

    writeln!(w, "[")?;
    for (i, r) in results.iter().enumerate() {
        let comma = if i < results.len() - 1 { "," } else { "" };
        writeln!(w, "  {{\"file\": {:?}, \"lines\": {}, \"dace_score\": {}, \"complexity\": {:?}}}{comma}",
            r.file, r.lines, r.dace_score, r.complexity)?;
    }
    writeln!(w, "]")?;

    Ok(())
}
